apps/emqx/src/emqx_broker.erl

```diff
@@ -518,32 +502,48 @@ handle_call({subscribe, Topic, SubPid, 0}, {From, _Tag}, State) ->
     Result = maybe_add_route(Existed, Topic, From),
     {reply, Result, State};
 handle_call({subscribe, Topic, SubPid, I}, {From, _Tag}, State) ->
-    Existed = ets:member(?SUBSCRIBER, Topic),
-    true = ets:insert(?SUBSCRIBER, [
-        {Topic, {shard, I}},
-        {{shard, Topic, I}, SubPid}
-    ]),
+    %% This will attempt to add a route per each new shard.
+    %% The overhead must be negligible, but the consistency in general
+    %% / race conditions safety is expected to be stronger.
+    Existed = ets:member(?SUBSCRIBER, {shard, Topic, I}),
+    Recs = [{{shard, Topic, I}, SubPid}],
+    Recs1 =
+        case Existed of
+            false -> [{Topic, {shard, I}} | Recs];
+            true -> Recs
+        end,
+    true = ets:insert(?SUBSCRIBER, Recs1),
     Result = maybe_add_route(Existed, Topic, From),
     {reply, Result, State};
 handle_call(Req, _From, State) ->
     ?SLOG(error, #{msg => "unexpected_call", call => Req}),
     {reply, ignored, State}.
 
+handle_cast({unsubscribed, Topic, shard, _I}, State) ->
+    _Result = maybe_delete_route(Topic),
```

Comment:
Note: `{t/#, 42}` can't send {delete, t/#}, only `{t/#, 0}` is allowed to do that. 

Let's imagine, `{t/#, 0}` broker sees that there are absolutely no subscribers (ets:member(?SUBSCRIBER, t/#) =:= true`) so it sends `{delete, t/#}` to route syncer. This indeed can be arrived in unspecified order with `{add, t/#}` from {t/#, 43}. 
But, if `{t/#, 43}` manages to send `{add, t/#}`, then there are must be some more subscribers for shard 0 that must be processed by `{t/#, 0}` broker.
 This is guareded by the atomic ETS sequence: if shard 43 is trying to add routes, than there are 42 shards behind. 
So, the final batch may look like:
```
{add, t/#},        % from {t/#, 43}
{delete, t/#}     % from {t,#, 0}
{add, t/#},        % from {t/#, 0} (can be several) 
```

Reply:
Well, it's probably still possible, though quite hard to imagine :sweat:
* shard `{t/#, 0}` deletes a route
* shard `{t/#, 43}` adds a route
These msgs  are re-ordered: `add, delete`. 
shard `{t/#, 0}` tries to add the route again (because if the sequence reached shard `{t/#, 43}`, than 0 shard must be full. 
However, none of the `{t/#, 0}` shard subscribers wins the race with `{t/#, 43}`, i,e, {t/#, 43} inserted a subscriber to ETS tab, and  `ets:member(?SUBSCRIBER, Topic)` returned `true` for all `{t/#, 0}` subscribers..

Reply:
> Note: `{t/#, 42}` can't send `{delete, t/#}`, only `{t/#, 0}` is allowed to do that.

Oops, sorry for the typo, I meant exactly that. 🫠

> shard `{t/#, 0}` tries to add the route again (because if the sequence reached shard `{t/#, 43}`, than 0 shard must be full.

Right. Thanks for the detailed explanation! I was under impression that counter updates are not atomic, and can stray away from the actual number of subscribers, assuming clients can be killed without notice. And it looks like it can overestimate that number AFAICT (though would be cool if you'll confirm or deny that).

But even with that in mind, I completely agree it's a very hard to reach scenario. Unsure if there are other scenarios with more processes / different reorderings with easier to reach inconsistencies.

> Perhaps we need to eliminate adding routes in shards other than 0 and cast a message to `pick({Topic, 0})` (similarly to how it worked before).

Wouldn't it defeat the purpose of this optimization? Or not too much?

Reply:
> Right. Thanks for the detailed explanation! I was under impression that counter updates are not atomic, and can stray away from the actual number of subscribers, assuming clients can be killed without notice. And it looks like it can overestimate that number AFAICT (though would be cool if you'll confirm or deny that).

You are right, clients can die fast and the sequence is reclaimed only eventually, but when clients die routes are not removed automatically. So, the sequence can indeed reflect a stale count of alive subscribers as well as doesn't immediately start allocating new subscribers to shard `{t/#, 0}`. 
But it doesn't seem dangerous for routes management :thinking:  If a number of subscribes die and new ones come, there is indeed no guarantee that at least one will go to shard `{t/#, 0}`, as sequence update is not immediate. 
However, when this happens once any new subscriber inserts itself into ?SUBSCRIBER and prevents the route from being removed. 

>Wouldn't it defeat the purpose of this optimization? Or not too much?

I think it shouldn't at all, we will cast one message per shard, something like:
```
handle_call({subscribe, Topic, SubPid, I}, {From, _Tag}, State) ->
    Existed = ets:member(?SUBSCRIBER, {shard, Topic, I}),
    Recs = [{{shard, Topic, I}, SubPid}],
    Recs1 =
        case Existed of
            false -> 
 +               cast(pick({Topic, 0}), {subscribe, Topic}),
                [{Topic, {shard, I}} | Recs];
            true -> 
                Recs
        end,
 -   true = ets:insert(?SUBSCRIBER, Recs1),
 -   Result = maybe_add_route(Existed, Topic, From),
    {reply, Result, State};
```

Reply:
> clients can die fast and the sequence is reclaimed only eventually

True. I was also thinking about the fact that `nextval` is called before inserting record into `?SUPOPTION` table, meaning that a client can bump the sequence and then die right away, so the sequence is never reclaimed. Extremely hard to achieve, although imaginable in a long-running system.

> I think it shouldn't at all, we will cast one message per shard, something like:

Nice! 💯 Thanks for the clarification, that should indeed exclude the possibility of any race conditions. This will probably need some polishing to preserve "call" semantics (i.e. reply when the route is propagated to the global table), but it seems easy to do.